### DISCUSSION TITLE
Undefined symbols compilation error when building in release configuration

### DISCUSSION THREAD

**Discussion by vladyslavsosiuk** [pointfreeco/swift-composable-architecture#2439] [General]
### Description

When there is a public reducer that uses an internal reducer in its body and compiling in release mode the compiler will throw an error `ld: Undefined symbols:`.
<img width="746" alt="image" src="https://github.com/pointfreeco/swift-composable-architecture/assets/24592433/2099d2d0-a497-4d6b-8679-7deeff82e0c2">
There is no error when compiling in Debug mode.
To workaround the issue simply mark the reducer public.

you can find a sample project here:
https://github.com/vladyslavsosiuk/CompilerBug
a link to the internal reducer that causes the error:
https://github.com/vladyslavsosiuk/CompilerBug/blob/d8215426ec7b6c7a5e8b8d897f393bb3080d9b2b/Modules/Sources/Child/Child.swift#L16

the `main` branch represents the issue
the `workaround` branch represents the workaround to get the project compiling in release configuration

### Checklist

- [ ] I have determined whether this bug is also reproducible in a vanilla SwiftUI project.
- [X] If possible, I've reproduced the issue using the `main` branch of this package.
- [X] This issue hasn't been addressed in an [existing GitHub issue](https://github.com/pointfreeco/swift-composable-architecture/issues) or [discussion](https://github.com/pointfreeco/swift-composable-architecture/discussions).

### Expected behavior

The app compiles without any errors

### Actual behavior

The app doesn't compile in Release configuration

### Steps to reproduce

1. Open the sample project
2. Make sure that Release configuration is selected by opening Scheme editor -> Build configuration -> Release
3. Compile the target
4. Observe the compilation error

### The Composable Architecture version information

15acb23

### Destination operating system

iOS 17, iOS 16.4

### Xcode version information

Xcode 14.3.1, 15 Beta 8

### Swift Compiler version information

```shell
swift-driver version: 1.87.1 Apple Swift version 5.9 (swiftlang-5.9.0.128.106 clang-1500.0.40.1)
Target: arm64-apple-macosx13.0
```

**Comment by mbrandonw** [comment]
Hi @vladyslavsosiuk, this appears to be a bug in the Swift compiler and not an issue in TCA. I'm moving this to a GitHub discussion, and feel free to continue discussing it there.

**Comment by lukeredpath** [comment]
Is this a compiler bug? I thought that the types of your internal reducers become part of the type signature for your public reducer’s body therefore the internal reducer types need to be public themselves. 

You can of course keep the internal reducer’s init function internal so the reducers cannot be used outside of that module.

**Reply by mamouneyya** [reply]
> therefore the internal reducer types _need_ to be public themselves.

But how would you explain being able to build in debug?
